staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -910,61 +913,46 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 		} else {
 			r.backoff.UpdateBackoff(r.URL(), err, resp.StatusCode)
 		}
-		if err != nil {
-			// "Connection reset by peer" or "apiserver is shutting down" are usually a transient errors.
-			// Thus in case of "GET" operations, we simply retry it.
-			// We are not automatically retrying "write" operations, as
-			// they are not idempotent.
-			if r.verb != "GET" {
-				return err
-			}
-			// For connection errors and apiserver shutdown errors retry.
-			if net.IsConnectionReset(err) || net.IsProbableEOF(err) {
-				// For the purpose of retry, we set the artificial "retry-after" response.
-				// TODO: Should we clean the original response if it exists?
-				resp = &http.Response{
-					StatusCode: http.StatusInternalServerError,
-					Header:     http.Header{"Retry-After": []string{"1"}},
-					Body:       ioutil.NopCloser(bytes.NewReader([]byte{})),
+
+		done := func() bool {
+			defer readAndCloseResponseBody(resp)
+
+			// if the the server returns an error in err, the response will be nil.
+			f := func(req *http.Request, resp *http.Response) {
+				if resp == nil {
+					return
 				}
-			} else {
-				return err
+				fn(req, resp)
 			}
-		}
 
-		done := func() bool {
-			// Ensure the response body is fully read and closed
-			// before we reconnect, so that we reuse the same TCP
-			// connection.
-			defer func() {
-				const maxBodySlurpSize = 2 << 10
-				if resp.ContentLength <= maxBodySlurpSize {
-					io.Copy(ioutil.Discard, &io.LimitedReader{R: resp.Body, N: maxBodySlurpSize})
+			var retry bool
+			retryAfter, retry = r.retry.NextRetry(req, resp, err, func(req *http.Request, err error) bool {
+				// "Connection reset by peer" or "apiserver is shutting down" are usually a transient errors.
+				// Thus in case of "GET" operations, we simply retry it.
+				// We are not automatically retrying "write" operations, as they are not idempotent.
+				if r.verb != "GET" {
+					return false
 				}
-				resp.Body.Close()
-			}()
-
-			retries++
```

Comment:
This change broke the threadsafety of `*Request`.  Previously, all methods on `*Request` tracked retry on function variables (thus different threads can call a request independently).  After this change, we can't.  I don't know that this is related to #109050, but we should not have made this change (we previously guaranteed the thread safety of methods called on Request).

Reply:
@smarterclayton - can you clarify this one? I think I'm not following this comment.

The Request itself doesn't seem to be claim to be thread-safe anywhere (and I'm not entirely sure what it would really mean to call different method requests from different threads). There isn't any locking anywhere here etc. either.

Maybe I missed something important in the review, but with this comment I still don't understand what that was.

Reply:
> we previously guaranteed the thread safety of methods called on Request

@smarterclayton  I am not sure how that would make sense semantically - multiple go routines access a `Request` instance concurrently, but i can make the `retry` a local variable with a factory function, I am working on a PR now.

Reply:
https://github.com/kubernetes/kubernetes/pull/109114

Reply:
You can call Stream() twice concurrently on a Request() and it would work before.  After this change, those two calls would compete IF there were retries.

Reply:
> You can call Stream() twice concurrently on a Request() and it would work before. After this change, those two calls would compete IF there were retries (you would have data races on the retry count).

dohh..
I'm not saying now that we are allowed to break it (agree with your we should fix it), but what would be the reason of doing that?

Reply:
Stream is a bit more questionable, reusing a LIST or GET request in some machinery as a closure would be more likely.